src/VisualStudio/Core/Def/DocumentOutline/DocumentSymbolDataViewModel.cs

```diff
@@ -0,0 +1,100 @@
+    {
+        public DocumentSymbolData Data { get; }
+        public ImmutableArray<DocumentSymbolDataViewModel> Children { get; }
+
+        /// <summary>
+        /// Necessary because we cannot convert to this type dynamically in WPF.
+        /// </summary>
+        public ImageMoniker ImageMoniker => Data.Glyph.GetImageMoniker();
+
+        private bool _isExpanded = true;
+        private bool _isSelected = false;
+
+        public bool IsExpanded
+        {
+            get => _isExpanded;
+            set => SetProperty(ref _isExpanded, value);
+        }
+
+        public bool IsSelected
+        {
+            get => _isSelected;
+            set => SetProperty(ref _isSelected, value);
+        }
+
+        public DocumentSymbolDataViewModel(
+            DocumentSymbolData data,
+            ImmutableArray<DocumentSymbolDataViewModel> children)
+        {
+            Data = data;
+            Children = children;
+        }
+
+        private static readonly PropertyChangedEventArgs _isExpandedPropertyChangedEventArgs = new PropertyChangedEventArgs(nameof(IsExpanded));
+        private static readonly PropertyChangedEventArgs _isSelectedPropertyChangedEventArgs = new PropertyChangedEventArgs(nameof(IsSelected));
+
+        private void NotifyPropertyChanged([CallerMemberName] string? propertyName = null)
+            => PropertyChanged?.Invoke(this, propertyName switch
+            {
+                nameof(IsExpanded) => _isExpandedPropertyChangedEventArgs,
+                nameof(IsSelected) => _isSelectedPropertyChangedEventArgs,
+                _ => new PropertyChangedEventArgs(propertyName)
+            });
+
+        public event PropertyChangedEventHandler? PropertyChanged;
+
+        private void SetProperty(ref bool field, bool value, [CallerMemberName] string propertyName = "")
+        {
+            // Note: we do not lock here. Worst case is that we fire multiple
+            //       NotifyPropertyChanged events which WPF can handle.
+            if (field == value)
+                return;
+
+            field = value;
+            NotifyPropertyChanged(propertyName);
+        }
+
+        public override bool Equals(object obj)
+            => Equals(obj as DocumentSymbolDataViewModel);
+
+        public bool Equals(DocumentSymbolDataViewModel? other)
+        {
+            if (other is null)
+                return false;
+
+            if (ReferenceEquals(this, other))
+                return true;
+
+            // If two view models are in the same location (across edits), we consider them the same.  We want to treat
+            // things like name edits as just mutating a model, not producing a new one.
+            var translatedRangeSpan = this.Data.RangeSpan.TranslateTo(other.Data.RangeSpan.Snapshot, SpanTrackingMode.EdgeInclusive);
+            return translatedRangeSpan == other.Data.RangeSpan;
```

Comment:
Do you want to use the entire range span or just the navigation span? I _think_ it'll be fine either way, but I wasn't sure if anything funky happened around before the start of a member (think XML doc comments) or after the end squiggle (think adding a new member after it.)

Reply:
that's a good question.  i think i'll leave as is, and we can try this out to see how well it works.  it's easy to revise if appropriate.